### PR TITLE
net: coap: Define default CoAP version

### PR DESCRIPTION
**Pull request by hubertmis**
RFC 7252 (CoAP) specifies value of the Version (Ver) field in the
protocol header to value 1. This patch defines default value of the
Version field to make packet initialization easier. All samples
and tests are updated to use the new COAP_VERSION field when
initializing a CoAP packet.

Signed-off-by: Hubert Miś <hubert.mis@nordicsemi.no>

### CODE REVIEW

File: subsys/net/lib/coap/coap_link_format.c [hunk 1]
```diff
@@ -653,7 +653,7 @@ int coap_well_known_core_get(struct coap_resource *resource,
 			     uint8_t *data, uint16_t len)
 {
 	struct coap_option query;
-	uint8_t token[8];
+	uint8_t token[COAP_TOKEN_MAX_LEN];
```
Comment (ceolin) [inline subsys/net/lib/coap/coap_link_format.c:656]: this change is not related with this commit, is it ?

Comment (hubertmis) [reply subsys/net/lib/coap/coap_link_format.c:656]: I've fixed a few other nits that I spotted near the commit area. Shall I move them to a separated commit?

**Comment by ceolin** [comment]
My suggestion is use a more descriptive name like `COAP_VERSION_01`, the rfc says:
` Implementations of this specification MUST set this field  to 1 (01 binary).  Other values are reserved for future versions.`

As our implementation allows to set the packet version it is better to have a specific identifier for this version. IMHO using `COAP_VERSION` means that there is only one version possible at time and consequently this could be hidden from public interface.

File: subsys/net/lib/coap/coap_link_format.c (hunk 1 above)
Comment (ceolin) [reply subsys/net/lib/coap/coap_link_format.c:656]: I think it is better

**Comment by hubertmis** [comment]
> IMHO using `COAP_VERSION` means that there is only one version possible at time and consequently this could be hidden from public interface.

I was thinking about it while replacing magic numbers with a define. The protocol header is designed in a way incorporating backward compatible changes without changing version number. It uses options to do that. If version number is ever changed, it would most probably mean that current packet initialization function parameters do not match the new protocol header. Together with adding new enumeration value `COAP_VERSION_2` we would probably add a function like `coap2_packet_init()` with other parameters.

If above assumptions are correct, it makes little sense to explicitly pass `COAP_VER_1` to each packet initialization sequence. We could simply remove this parameter from public API and assume value 1.

**Comment by rlubos** [comment]
> > IMHO using `COAP_VERSION` means that there is only one version possible at time and consequently this could be hidden from public interface.
> 
> I was thinking about it while replacing magic numbers with a define. The protocol header is designed in a way incorporating backward compatible changes without changing version number. It uses options to do that. If version number is ever changed, it would **most probably** mean that current packet initialization function parameters do not match the new protocol header. Together with adding new enumeration value `COAP_VERSION_2` we would **probably** add a function like `coap2_packet_init()` with other parameters.
> 
> If above assumptions are correct, it makes little sense to explicitly pass `COAP_VER_1` to each packet initialization sequence. We could simply remove this parameter from public API and assume value 1.

This reasoning makes sense, but we're predicting the future  a bit here. For now, our CoAP API is rather low-level (exposes a lot of details to the user) and the protocol version is one of these details. Personally, I'd rather avoid API changes unless necessary (or otherwise makes a reasonable improvement). If it turns out that any potential future CoAP specification is not compatible with the API, we can consider changing it. 

So personally, I'd stick to specifying a constant for `COAP_VERSION_01` in this PR. But I have no strong objections if there are more voices for `version` parameter removal from the API, and using a predefined value.

**Comment by jukkar** [comment]
>  I'd stick to specifying a constant for `COAP_VERSION_01` in this PR. But I have no strong objections if there are more voices for `version` parameter removal from the API, and using a predefined value.

I have no preference, Robert proposal sounds good.

**Comment by nandojve** [comment]
+1 to Robert proposal

**Comment by hubertmis** [comment]
@rlubos, well argued. The only modification I'll add to your proposal is to use decimal (`COAP_VERSION_1`) instead of binary version number (`COAP_VERSION_01`). Decimal numbers are easier to understand

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by nandojve — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

File: include/net/coap.h [hunk 2]
```diff
@@ -77,6 +77,8 @@ enum coap_method {
 
 #define COAP_REQUEST_MASK 0x07
 
+#define COAP_VERSION_1 1
```
Comment (ceolin) [inline include/net/coap.h:80]: nitpick, uses `1U` since it is used as unsigned.

```
#define COAP_VERSION_1 1U
```
